README.pod
=pod

=head1 NAME

Date::Easy - easy dates with Time::Piece compatibility

=head1 VERSION

This document describes version 0.03_01 of Date::Easy.

=head1 SYNOPSIS

    use Date::Easy


    # DATES

    # guaranteed to have a time of midnight
    my $d = date("3-Sep-1940");

    # addition and subtraction work in increments of days
    my $tomorrow = today + 1;
    my $last_week = today - 7;


    # DATETIMES

    # default timezone is your local zone
    my $dt = datetime("3/31/2012 7:38am");

    # addition and subtraction work in increments of seconds
    my $this_time_yesterday = now - 60*60*24;
    my $after_30_minutes = now + 30 * 60;

    # if you prefer UTC
    my $utc = datetime(UTC => "2016-03-07 01:22:16PST-0800");

    # or UTC for all your objects
    use Date::Easy 'UTC';
    say datetime("Jan 1 2000 midnight")->time_zone;
    # prints "UTC"


    # UNITS

    # basically just like the interface from Date::Piece
    say today + 3 * months;
    say now + 2 * hours - 5 * minutes;
    say 8 * weeks;
    # prints "8 weeks"

=head1 DESCRIPTION

Date::Easy provides simple date and datetime objects that will do what you expect, provided you
expect them to do the right things.  At its heart, a C<use Date::Easy> statement is just a shortcut
for this:

    use Date::Easy::Date ':all';
    use Date::Easy::Datetime ':all';
    use Date::Easy::Units ':all';

So, for full details, you should see the docs for L<Date::Easy::Date>, L<Date::Easy::Datetime> and
L<Date::Easy::Units>.  However, there are also a few parameters you can pass to Date::Easy (see
L</USAGE>).

=head2 Quick Start

A "datetime" (Date::Easy::Datetime) is an object which represents a date and time (internally, it's
just a L<Time::Piece> object, which is, at its heart, just a number of epoch seconds).  A "date"
(Date::Easy::Date) is just a datetime whose time portion is always guaranteed to be midnight (and is
therefore irrelevant).  When you C<use Date::Easy>, you get two ways to create dates:

    my $t = today;
    my $d = date($human_readable_string);

and two ways to create datetimes:

    my $n = now;
    my $dt = datetime($human_readable_string);

which pretty much do exactly what you think they do.  Once you have them, you can access individual
attributes of the objects:

    say "day is ", $d->day;
    say "hour is ", $dt->hour;

or you can do simple date math with them.  You may add and subtract integers to/from a date, which
are interpreted as days, and you may add and subtract integers to/from a datetime, which are
interpreted as seconds.  Adding and subtracting units objects to/from dates and datetimes add or
subtract the appropriate quantity of the appropriate unit to/from the date or datetime.

That's really all there is to it, for the basics.

=head2 Limitations

=head3 Time Zones

A date object is always in UTC.  When a string is parsed to get a date, any timezone information in
that string is ignored.  This avoids surprising results such as C<date("20-Jun-2016 9:00pm PDT")>
turning into June 21.

A datetime object is by default in your local timezone (whatever that is).  You can force it to be
in UTC instead in a number of different ways (just search for "UTC" on this page and also on the
page for L<Date::Easy::Datetime>.

=head3 Minima and Maxima

The minimum dates and datetimes that you can represent using Date::Easy objects are the same that
can be represented by epoch seconds (to be more precise, they are the same that can be accepted by
L<Time::Local>'s C<timegm> and C<timelocal>).  For 64-bit machines, I know (from experimentation)
this range to be 1-Jan-1000 00:00:00 to 31-Dec-2899 23:59:59.  For 32-bit machines, I I<believe> it
to be 13-Dec-1901 20:45:52 to 19-Jan-2038 03:14:07, but only prior to Perl 5.12 (Perl 5.12 and above
should handle your epoch seconds as a 64-bit int even when the underlying architecture is 32-bit).

If you are are using a Perl version before 5.12 and your underlying C<time_t> is represented as an
I<unsigned> integer, then all bets are off for you.

=head3 String Formats

The human-readable formats understood by C<date> and C<datetime> are the union of those understood
by L<Date::Parse> and L<Time::ParseDate>.  Date::Parse is tried first, except for a few minor
optimizations where it's easy to know in advance that it can't possibly recognize the format.

=head3 Daylight Savings Time

Date::Easy knows exactly as much about DST as C<localtime>, C<gmtime>, and L<Time::Local> do, which
is to say, it will probably handle most common uses, but may fail for pathological cases.

=head3 Leap Seconds

Date::Easy doesn't deal with leap seconds at all.  In fact, it's likely that the unit tests will
fail during module install if your local timezone includes leap seconds.

=head3 Language

Currently, Date::Easy only speaks English.  Specifically, that means:

=over

=item *

When parsing human-readable strings, it can only understand abbreviations and whole names of days of
the week and months of the year if they are in English.

=item *

When converting to human-readable strings, using C<strftime> (or C<as> with a non-class-string
argument, which just calls C<strftime> underneath), it will I<probably> only render days of the week
and months of the year in English.  However, it I<may> respect the current locale, depending on your
system's underlying POSIX C<strftime> implementation.

=item *

If you convert a unit (L<Date::Easy::Units>) to a string, you get the English name for the unit.

=back

=head1 USAGE

There are a few parameters you can pass to Date::Easy at C<use> time.  These are passed through to
L<Date::Easy::Datetime>.  Thus the following are equivalent:

    use Date::Easy 'local';
    # is the same as:
    use Date::Easy::Date ':all';
    use Date::Easy::Datetime qw< :all local >;
    use Date::Easy::Units ':all';

    use Date::Easy 'UTC';
    # is the same as:
    use Date::Easy::Date ':all';
    use Date::Easy::Datetime qw< :all UTC >;
    use Date::Easy::Units ':all';

    use Date::Easy 'GMT';
    # is the same as:
    use Date::Easy::Date ':all';
    use Date::Easy::Datetime qw< :all GMT >;
    use Date::Easy::Units ':all';

As the Date::Easy::Datetime docs will tell you, "UTC" and "GMT" are exactly equivalent as far as
Date::Easy is concerned.  Passing "local" is redundant, as it is the default, but perhaps you just
want to be explicit.

=head1 DETAILS

For more details on datetimes, see L<Date::Easy::Datetime>.

For more details on dates, see L<Date::Easy::Date>.

For more details on units, see L<Date::Easy::Units>.

=head1 INSPIRATION AND CREDITS

For far more than you ever wanted to know about Date::Easy, including inspirations and design goals,
please refer to my blog series I<L<A Date with
CPAN|http://blogs.perl.org/users/buddy_burden/2015/09/a-date-with-cpan-part-1-state-of-the-union.html>>.

The implementation of both dates and datetimes is almost entirely handled by L<Time::Piece>, by Matt
Sergeant and Jarkko Hietaniemi (based on ideas from Larry Wall).

The interface of datetimes (such as method names and data ranges) more closely conforms to that of
L<DateTime>, by Dave Rolsky.

The interface of constructors (such as `today` and `date`) and that of units objects is shamelessly
stolen from L<Date::Piece>, by Eric Wilhelm.

Date::Easy exists by standing on the shoulders of these giants.  All I did was glue the bits
together.

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Perldoc

You can find documentation for this module with the perldoc command.

  perldoc Date::Easy

=head2 Bugs / Feature Requests

This module is on GitHub.  Feel free to fork and submit patches.  Please note that I develop
via TDD (Test-Driven Development), so a patch that includes a failing test is much more
likely to get accepted (or at least likely to get accepted more quickly).

If you just want to report a problem or suggest a feature, that's okay too.  You can create
an issue on GitHub here: L<http://github.com/barefootcoder/date-easy/issues>.

=head2 Source Code

none
L<https://github.com/barefootcoder/date-easy>

  git clone https://github.com/barefootcoder/date-easy.git

=head1 AUTHOR

Buddy Burden <barefootcoder@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2017 by Buddy Burden.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


Changes
Revision history for Date-Easy

{{$NEXT}}

0.10      2020-01-14 13:57:52-08:00 America/Los_Angeles

    [Distro Fixes]
    * Fix failing unit test caused by imperfect mirroring of `gmtime`/`timegm`
      github #14 (thx Slaven Rezić (SREZIC))
    [Docs]
    * Fix a small typo

0.09      2019-09-03 16:34:03-07:00 America/Los_Angeles

    [Docs]
    * Reviewed all POD and fixed a few typos
    * Fixed a POD link (github #13--thx manwar)

    [Distro Fixes]
    * Fix format test so it can handle weird chars in default format
      github #9 (thx richlv)

0.08      2019-06-28 17:42:42-07:00 America/Los_Angeles

    [Efficiency Improvements]
    * resimplify `strftime` implementation to simple delegation
      this made possible by the fixing of Time::Piece GitHub #24 (thx ESAYM!)
      that fix means we must require at least 1.32 of Time::Piece

    [Distro Fixes]
    * Fix Kwalitee issues
      github #11 (thx Zak B. Elep (ZAKAME))
    * Fix failing unit test caused by bug in `str2time`
      github #12 (thx Slaven Rezić (SREZIC))
    * Correct broken distro (no `provides` in META.*)

0.07      2019-04-16 00:33:10-07:00 America/Los_Angeles

    -- broken distro --

0.06      2018-06-15 22:34:29-07:00 America/Los_Angeles

    [Bug Fixes]
    * workaround for upstream bug in Date::Parse [base bug RT/53413]
      now doing a more faithful recreation of D::P's strptime
      and using it for both dates and datetimes

0.05      2018-05-31 02:50:30-07:00 America/Los_Angeles

    [New Features]
    * Can now subtract one datetime from another (gives seconds).
    * Can now subtract one date from another (gives days).

    [Distro Fixes]
    * Change date-parse unit tests to calculate epoch dates
      so they work in all timezones.
      github #10 (thx Ron Savage (RSAVAGE))

0.04      2017-07-27 17:47:54-07:00 America/Los_Angeles

    [Bug Fixes]
    * `add_months` was returning a Time::Piece instead of a Date::Easy
      object (fixed)

    [New Features]
    * New `day_of_year` method
      returns day of year (1-366)
    * New `split` method
      splits to (Y, m, d) for date, (Y, m, d, H, M, S) for datetime
    * New methods: `add_seconds`, `subtract_seconds`, `add_minutes`,
      `subtract_minutes`, `add_hours`, `subtract_hours`, `add_days`,
      `subtract_days`, `add_weeks`, `subtract_weeks`, `subtract_months`,
      `add_years`, `subtract_years`
    * New class: Date::Easy::Units
      simpler interface for date math

    [Docs]
    * Added "Language" section to "Limitations"
    * Added "Inspiration and Credits" section to POD
    * Documented effects of upstream bug RT/105031

0.03      2017-02-21 19:45:46-08:00 America/Los_Angeles

    [New Features]
    * New `add_months` method
    * New `iso8601` method (and `iso` alias)

    [Interface Changes]
    * Have `as` able to recognize a string like '-Ymd'
      (same as `strftime('%Y-%m-%d')`)

    [Docs]
    * Make all method headers consistent (show arguments)

0.02      2016-06-25 02:00:48-07:00 America/Los_Angeles

    [New Features]
    * New `time_zone` method

    [Interface Changes]
    * Make `add` and `subtract` in ::Date private methods
      (this fixes POD coverage tests)

    [Distro Fixes]
    * Stop using `\h` in test regexen because 5.8 doesn't recognize it

    [Docs]
    * Added basic POD

0.01_03   2016-05-25 22:43:31PDT-0700 America/Los_Angeles (TRIAL RELEASE)

    [Bug Fixes]
    * The "%s" format specifier now always returns the same value as the
      epoch method, regardless of what the underlying implementation thinks

    [Distro Fixes]
    * Fixed `strftime` unit tests being locked to the author's timezone
      github #6 (thx Slaven Rezić (SREZIC))
    * Removed check for "%p" in `strftime` test because it's locale-specific
      github #7 (thx Slaven Rezić (SREZIC))

0.01_02   2016-05-22 01:48:55PDT-0700 America/Los_Angeles (TRIAL RELEASE)

    [Interface Changes]
    * ::Date and ::Datetime now have accessors that follow the UI of
      DateTime rather than Time::Piece
    * ::Date and ::Datetime no longer inherit from Time::Piece

    [Distro Fixes]
    * Fixed multiple cases of datetimes not matching due to clock rolling over
      github #4 (thx Slaven Rezić (SREZIC))
    * Fixed out-of-range tests on 32-bit machines
      github #3 (thx Slaven Rezić (SREZIC))
    * Fixed bug where List::Util might be too low a version during testing
      github #5 (thx Slaven Rezić (SREZIC))

0.01_01   2016-04-16 18:52:35-07:00 America/Los_Angeles (TRIAL RELEASE)

    [Bug Fixes]
    * Correct the bug whereby handing off to Time::ParseDate::parsedate
      used the wrong DST flag

    [Distro Fixes]
    * Incorrect issues link fixed
    * Now prints timezone info to help identify smoker problems
      (thx Andreas Koenig (ANDK)!)
    * Added author test for parsing dates in all timezones
      (should provide better test coverage on release)
    * Marked bogusly failing datetime test as TODO
      (see also: https://github.com/muir/Time-modules/issues/8)

0.01      2016-03-07 01:22:16PST-0800 America/Los_Angeles

    * initial version


t/lib/DateParseTests.pm
package DateParseTests;

use strict;
use warnings;
use autodie;

# This giant variable is shamlessly stolen from the test file for Date::Parse.
# I've reproduced the comments as well to preserve the authorship credits.
# This is taken from GBARR/TimeDate-2.30/t/getdate.t.
# (But the code hasn't changed since TimeDate-1.17.)
# Mild reformatting applied.


#Thanks to Andreas Koenig for converting all those dates to numbers
#and adding the folloing acknowledgement into Date/t/getdate.t
#Thanks to Graham Barr for writing these tests. Slightly adjusted for
#the C version by Andreas Koenig, 96-06-08.

my $data = qq!1995-01-24             ;790905600
1995-06-24                           ;803952000
92/01/02 12:01                       ;694353660
92/01/02 12:01 AM                    ;694310460
92/01/02 12:01 PM                    ;694353660
2002-11-02 11pm GMT                  ;1036278000
2002-11-02 11 pm GMT                 ;1036278000
1995-01-24  GMT                      ;790905600
1995-01-24  BST                      ;790902000
1995-06-24  GMT                      ;803952000
1995-06-24  BST                      ;803948400
1992-1-2 12:01 GMT                   ;694353660
1992-1-2-12:01:00 GMT                ;694353660
Wed, 16 Jun 94 07:29:35 CST          ;771773375
Wed,16 Jun 94 07:29:35 CST           ;771773375
Wed, 16 Nov 94 07:29:35 CST          ;784992575
Mon, 21 Nov 94 07:42:23 CST          ;785425343
Mon, 21 Nov 94 04:28:18 CST          ;785413698
Tue, 15 Nov 94 09:15:10 GMT          ;784890910
Wed, 16 Nov 94 09:39:49 GMT          ;784978789
Wed, 16 Nov 94 09:23:17 GMT          ;784977797
Wed, 16 Nov 94 12:39:49 GMT          ;784989589
Wed, 16 Nov 94 14:03:06 GMT          ;784994586
Wed, 16 Nov 94 05:30:51 CST          ;784985451
Thu, 17 Nov 94 03:19:30 CST          ;785063970
1994:11:21T14:05:32+0000             ;785426732
Mon, 21 Nov 94 14:05:32 GMT          ;785426732
Mon, 14 Nov 94 15:08:49 CST          ;784847329
Wed, 16 Nov 94 14:48:06 GMT          ;784997286
Thu, 17 Nov 94 14:22:03 GMT          ;785082123
Wed, 16 Nov 94 14:36:00 GMT          ;784996560
Wed, 16 Nov 94 09:23:17 GMT          ;784977797
Wed, 16 Nov 94 10:01:43 GMT          ;784980103
Wed, 16 Nov 94 15:03:35 GMT          ;784998215
Mon, 21 Nov 94 13:55:19 GMT          ;785426119
Wed, 16 Nov 94 08:46:11 CST          ;784997171
Wed, 9 Nov 1994 09:50:32 -0500 (EST) ;784392632
Thu, 13 Oct 94 10:13:13 -0700        ;782068393
Sat, 19 Nov 1994 16:59:14 +0100      ;785260754
Thu, 3 Nov 94 14:10:47 EST           ;783889847
Thu, 3 Nov 94 21:51:09 EST           ;783917469
Fri, 4 Nov 94 9:24:52 EST            ;783959092
Wed, 9 Nov 94 09:38:54 EST           ;784391934
Mon, 14 Nov 94 13:20:12 EST          ;784837212
Wed, 16 Nov 94 17:09:13 EST          ;785023753
Tue, 15 Nov 94 12:27:01 PST          ;784931221
Fri, 18 Nov 1994 07:34:05 -0600      ;785165645
Mon, 21 Nov 94 14:34:28 -0500        ;785446468
Fri, 18 Nov 1994 12:05:47 -0800 (PST);785189147
Fri, 18 Nov 1994 12:36:26 -0800 (PST);785190986
Wed, 16 Nov 1994 15:58:58 GMT        ;785001538
Sun, 06 Nov 94 14:27:40 -0500        ;784150060
Mon, 07 Nov 94 08:20:13 -0500        ;784214413
Mon, 07 Nov 94 16:48:42 -0500        ;784244922
Wed, 09 Nov 94 15:46:16 -0500        ;784413976
Sun, 6 Nov 1994 02:38:17 -0800       ;784118297
Tue, 1 Nov 1994 13:53:49 -0500       ;783716029
Tue, 15 Nov 94 08:31:59 +0100        ;784884719
Sun, 6 Nov 1994 11:09:12 -0500 (IST) ;784138152
Fri, 4 Nov 94 12:52:10 EST           ;783971530
Mon, 31 Oct 1994 14:17:39 -0500 (EST);783631059
Mon, 14 Nov 94 11:25:00 CST          ;784833900
Mon, 14 Nov 94 13:26:29 CST          ;784841189
Fri, 18 Nov 94 8:42:47 CST           ;785169767
Thu, 17 Nov 94 14:32:01 +0900        ;785050321
Wed, 2 Nov 94 18:16:31 +0100         ;783796591
Fri, 18 Nov 94 10:46:26 +0100        ;785151986
Tue, 8 Nov 1994 22:39:28 +0200       ;784327168
Wed, 16 Nov 1994 10:01:08 -0500 (EST);784998068
Wed, 2 Nov 1994 16:59:42 -0800       ;783824382
Wed, 9 Nov 94 10:00:23 PST           ;784404023
Fri, 18 Nov 94 17:01:43 PST          ;785206903
Mon, 14 Nov 1994 14:47:46 -0500      ;784842466
Mon, 21 Nov 1994 04:56:04 -0500 (EST);785411764
Mon, 21 Nov 1994 11:50:12 -0800      ;785447412
Sat, 5 Nov 1994 14:04:16 -0600 (CST) ;784065856
Sat, 05 Nov 94 13:10:13 MST          ;784066213
Wed, 02 Nov 94 10:47:48 -0800        ;783802068
Wed, 02 Nov 94 13:19:15 -0800        ;783811155
Thu, 03 Nov 94 15:27:07 -0800        ;783905227
Fri, 04 Nov 94 09:12:12 -0800        ;783969132
Wed, 9 Nov 1994 10:13:03 +0000 (GMT) ;784375983
Wed, 9 Nov 1994 15:28:37 +0000 (GMT) ;784394917
Wed, 2 Nov 1994 17:37:41 +0100 (MET) ;783794261
05 Nov 94 14:22:19 PST               ;784074139
16 Nov 94 22:28:20 PST               ;785053700
Tue, 1 Nov 1994 19:51:15 -0800       ;783748275
Wed, 2 Nov 94 12:21:23 GMT           ;783778883
Fri, 18 Nov 94 18:07:03 GMT          ;785182023
Wed, 16 Nov 1994 11:26:27 -0500      ;785003187
Sun, 6 Nov 1994 13:48:49 -0500       ;784147729
Tue, 8 Nov 1994 13:19:37 -0800       ;784329577
Fri, 18 Nov 1994 11:01:12 -0800      ;785185272
Mon, 21 Nov 1994 00:47:58 -0500      ;785396878
Mon, 7 Nov 1994 14:22:48 -0800 (PST) ;784246968
Wed, 16 Nov 1994 15:56:45 -0800 (PST);785030205
Thu, 3 Nov 1994 13:17:47 +0000       ;783868667
Wed, 9 Nov 1994 17:32:50 -0500 (EST) ;784420370
Wed, 9 Nov 94 16:31:52 PST           ;784427512
Wed, 09 Nov 94 10:41:10 -0800        ;784406470
Wed, 9 Nov 94 08:42:22 MST           ;784395742
Mon, 14 Nov 1994 08:32:13 -0800      ;784830733
Mon, 14 Nov 1994 11:34:32 -0500 (EST);784830872
Mon, 14 Nov 94 16:48:09 GMT          ;784831689
Tue, 15 Nov 1994 10:27:33 +0000      ;784895253
Wed, 02 Nov 94 13:56:54 MST          ;783809814
Thu, 03 Nov 94 15:24:45 MST          ;783901485
Thu, 3 Nov 1994 15:13:53 -0700 (MST) ;783900833
Fri, 04 Nov 94 08:15:13 MST          ;783962113
Thu, 3 Nov 94 18:15:47 EST           ;783904547
Tue, 08 Nov 94 07:02:33 MST          ;784303353
Thu, 3 Nov 94 18:15:47 EST           ;783904547
Tue, 15 Nov 94 07:26:05 MST          ;784909565
Wed, 2 Nov 1994 00:00:55 -0600 (CST) ;783756055
Sun, 6 Nov 1994 01:19:13 -0600 (CST) ;784106353
Mon, 7 Nov 1994 23:16:57 -0600 (CST) ;784271817
Tue, 08 Nov 1994 13:21:21 -0600      ;784322481
Mon, 07 Nov 94 13:47:37 PST          ;784244857
Tue, 08 Nov 94 11:23:19 PST          ;784322599
Tue, 01 Nov 1994 11:28:25 -0800      ;783718105
Tue, 15 Nov 1994 13:11:47 -0800      ;784933907
Tue, 15 Nov 1994 13:18:38 -0800      ;784934318
Tue, 15 Nov 1994 0:18:38 -0800       ;784887518
Jul 13 1999 1:23P GMT                ;931872180
Jul 13 1999 1:23P.M GMT              ;931872180
Jul 13 1999 1:23P.M. GMT             ;931872180
Jul-13-1999 1:23P.M GMT              ;931872180
Jul/13/1999 1:23P.M. GMT             ;931872180
2001-02-26T13:44:12-0700             ;983220252
2001-02-26T13:44-0700                ;983220240
2001-02-26T13-0700                   ;983217600
2001-02-26T13-44-12-0700             ;983220252
2001-02-26T13-44-0700                ;983220240
2001-02-26T13-0700                   ;983217600
07/Nov/2000:16:45:56 +0100           ;973611956
2002-11-07T23:31:49-05:00            ;1036729909
2003-02-17T07:45:56Z                 ;1045467956
2003-02-17T08:14:07.198189+0000      ;1045469647.198189
20020722T100000Z                     ;1027332000
Jul 22 10:00:00 UTC 2002             ;1027332000
2002-07-22 10:00:00Z                 ;1027332000
2002-07-22 10:00:00 Z                ;1027332000
2002-07-22 10:00 Z                   ;1027332000
2002-07-22 10:00Z                    ;1027332000
2002-07-22 10:00 +100                ;1027328400
2002-07-22 10:00 +0100               ;1027328400
!;

# END stolen code


# Now let's turn this data into a simple hash:

our %DATE_PARSE_TESTS =
	map { split(/\s*;/) }
	split("\n", $data);


# And we'll go ahead and export our hash.
# This is only ever used by test files anyway.

use parent 'Exporter';
our @EXPORT_OK = qw< %DATE_PARSE_TESTS _date_parse_has_timezone _date_parse_remove_timezone >;


# Some subs to help us deal with timezones.
# These are exported too.

my $TIMEZONE_REGEX = qr/ ( [+-] \d{1,2} :? \d{2} | [A-Z]{3} \h* ( \d{4} )? | [+-] \d{4} \h \( [A-Z]{3} \) | Z ) $/x;

sub _date_parse_has_timezone
{
	return shift =~ /$TIMEZONE_REGEX/;
}

sub _date_parse_remove_timezone
{
	my $time = shift;

	no warnings 'uninitialized';					# because $2 will often be undefined
	$time =~ s/$TIMEZONE_REGEX/$2/;

	return $time;
}


1;


lib/Date/Easy.pm
package Date::Easy;

use strict;
use warnings;
use autodie;

# VERSION

use Date::Easy::Date ();
use Date::Easy::Datetime ();
use Date::Easy::Units ();

use Exporter;
use parent 'Exporter';
our @EXPORT = ( @Date::Easy::Date::EXPORT_OK, @Date::Easy::Datetime::EXPORT_OK, @Date::Easy::Units::EXPORT_OK, );


sub import
{
	Date::Easy::Date->import(':all');
	Date::Easy::Datetime->import(':all', @_[1..$#_]);
	Date::Easy::Units->import(':all');
	@_ = (shift);									# throw away all args except the first (package name)
	goto &Exporter::import;
}



1;



# ABSTRACT: easy dates with Time::Piece compatibility
# COPYRIGHT


=head1 SYNOPSIS

    use Date::Easy


    # DATES

    # guaranteed to have a time of midnight
    my $d = date("3-Sep-1940");

    # addition and subtraction work in increments of days
    my $tomorrow = today + 1;
    my $last_week = today - 7;


    # DATETIMES

    # default timezone is your local zone
    my $dt = datetime("3/31/2012 7:38am");

    # addition and subtraction work in increments of seconds
    my $this_time_yesterday = now - 60*60*24;
    my $after_30_minutes = now + 30 * 60;

    # if you prefer UTC
    my $utc = datetime(UTC => "2016-03-07 01:22:16PST-0800");

    # or UTC for all your objects
    use Date::Easy 'UTC';
    say datetime("Jan 1 2000 midnight")->time_zone;
    # prints "UTC"


    # UNITS

    # basically just like the interface from Date::Piece
    say today + 3 * months;
    say now + 2 * hours - 5 * minutes;
    say 8 * weeks;
    # prints "8 weeks"


=head1 DESCRIPTION

Date::Easy provides simple date and datetime objects that will do what you expect, provided you
expect them to do the right things.  At its heart, a C<use Date::Easy> statement is just a shortcut
for this:

    use Date::Easy::Date ':all';
    use Date::Easy::Datetime ':all';
    use Date::Easy::Units ':all';

So, for full details, you should see the docs for L<Date::Easy::Date>, L<Date::Easy::Datetime> and
L<Date::Easy::Units>.  However, there are also a few parameters you can pass to Date::Easy (see
L</USAGE>).

=head2 Quick Start

A "datetime" (Date::Easy::Datetime) is an object which represents a date and time (internally, it's
just a L<Time::Piece> object, which is, at its heart, just a number of epoch seconds).  A "date"
(Date::Easy::Date) is just a datetime whose time portion is always guaranteed to be midnight (and is
therefore irrelevant).  When you C<use Date::Easy>, you get two ways to create dates:

    my $t = today;
    my $d = date($human_readable_string);

and two ways to create datetimes:

    my $n = now;
    my $dt = datetime($human_readable_string);

which pretty much do exactly what you think they do.  Once you have them, you can access individual
attributes of the objects:

    say "day is ", $d->day;
    say "hour is ", $dt->hour;

or you can do simple date math with them.  You may add and subtract integers to/from a date, which
are interpreted as days, and you may add and subtract integers to/from a datetime, which are
interpreted as seconds.  Adding and subtracting units objects to/from dates and datetimes add or
subtract the appropriate quantity of the appropriate unit to/from the date or datetime.

That's really all there is to it, for the basics.

=head2 Limitations

=head3 Time Zones

A date object is always in UTC.  When a string is parsed to get a date, any timezone information in
that string is ignored.  This avoids surprising results such as C<date("20-Jun-2016 9:00pm PDT")>
turning into June 21.

A datetime object is by default in your local timezone (whatever that is).  You can force it to be
in UTC instead in a number of different ways (just search for "UTC" on this page and also on the
page for L<Date::Easy::Datetime>.

=head3 Minima and Maxima

The minimum dates and datetimes that you can represent using Date::Easy objects are the same that
can be represented by epoch seconds (to be more precise, they are the same that can be accepted by
L<Time::Local>'s C<timegm> and C<timelocal>).  For 64-bit machines, I know (from experimentation)
this range to be 1-Jan-1000 00:00:00 to 31-Dec-2899 23:59:59.  For 32-bit machines, I I<believe> it
to be 13-Dec-1901 20:45:52 to 19-Jan-2038 03:14:07, but only prior to Perl 5.12 (Perl 5.12 and above
should handle your epoch seconds as a 64-bit int even when the underlying architecture is 32-bit).

If you are are using a Perl version before 5.12 and your underlying C<time_t> is represented as an
I<unsigned> integer, then all bets are off for you.

=head3 String Formats

The human-readable formats understood by C<date> and C<datetime> are the union of those understood
by L<Date::Parse> and L<Time::ParseDate>.  Date::Parse is tried first, except for a few minor
optimizations where it's easy to know in advance that it can't possibly recognize the format.

=head3 Daylight Savings Time

Date::Easy knows exactly as much about DST as C<localtime>, C<gmtime>, and L<Time::Local> do, which
is to say, it will probably handle most common uses, but may fail for pathological cases.

=head3 Leap Seconds

Date::Easy doesn't deal with leap seconds at all.  In fact, it's likely that the unit tests will
fail during module install if your local timezone includes leap seconds.

=head3 Language

Currently, Date::Easy only speaks English.  Specifically, that means:

=over

=item *

When parsing human-readable strings, it can only understand abbreviations and whole names of days of
the week and months of the year if they are in English.

=item *

When converting to human-readable strings, using C<strftime> (or C<as> with a non-class-string
argument, which just calls C<strftime> underneath), it will I<probably> only render days of the week
and months of the year in English.  However, it I<may> respect the current locale, depending on your
system's underlying POSIX C<strftime> implementation.

=item *

If you convert a unit (L<Date::Easy::Units>) to a string, you get the English name for the unit.

=back


=head1 USAGE

There are a few parameters you can pass to Date::Easy at C<use> time.  These are passed through to
L<Date::Easy::Datetime>.  Thus the following are equivalent:

    use Date::Easy 'local';
    # is the same as:
    use Date::Easy::Date ':all';
    use Date::Easy::Datetime qw< :all local >;
    use Date::Easy::Units ':all';

    use Date::Easy 'UTC';
    # is the same as:
    use Date::Easy::Date ':all';
    use Date::Easy::Datetime qw< :all UTC >;
    use Date::Easy::Units ':all';

    use Date::Easy 'GMT';
    # is the same as:
    use Date::Easy::Date ':all';
    use Date::Easy::Datetime qw< :all GMT >;
    use Date::Easy::Units ':all';

As the Date::Easy::Datetime docs will tell you, "UTC" and "GMT" are exactly equivalent as far as
Date::Easy is concerned.  Passing "local" is redundant, as it is the default, but perhaps you just
want to be explicit.


=head1 DETAILS

For more details on datetimes, see L<Date::Easy::Datetime>.

For more details on dates, see L<Date::Easy::Date>.

For more details on units, see L<Date::Easy::Units>.


=head1 INSPIRATION AND CREDITS

For far more than you ever wanted to know about Date::Easy, including inspirations and design goals,
please refer to my blog series I<L<A Date with
CPAN|http://blogs.perl.org/users/buddy_burden/2015/09/a-date-with-cpan-part-1-state-of-the-union.html>>.

The implementation of both dates and datetimes is almost entirely handled by L<Time::Piece>, by Matt
Sergeant and Jarkko Hietaniemi (based on ideas from Larry Wall).

The interface of datetimes (such as method names and data ranges) more closely conforms to that of
L<DateTime>, by Dave Rolsky.

The interface of constructors (such as `today` and `date`) and that of units objects is shamelessly
stolen from L<Date::Piece>, by Eric Wilhelm.

Date::Easy exists by standing on the shoulders of these giants.  All I did was glue the bits
together.
